library/src/compute/data.rs

```diff
@@ -210,6 +210,99 @@ fn format_json_error(error: serde_json::Error) -> String {
 }
 
+/// Read structured data from a YAML file.
+///
+/// The file must contain a valid YAML object or array. YAML mappings will be
+/// converted into Typst dictionaries, and YAML sequences will be converted into
+/// Typst arrays. Strings and booleans will be converted into the Typst
+/// equivalents, null-values (`null`, `~` or empty ``) will be converted into 
+/// `{none}`, and numbers will be converted to floats or integers depending on 
+/// whether they are whole numbers.
+///
+/// Note that mapping keys that are not a string, a number, a boolean or a 
+/// null-value will cause the entry to be discarded. Otherwise it will be 
+/// converted to a string.
+/// Custom YAML tags are ignored, though the loaderd value will still be 
+/// present.
+/// 
+/// The function returns a dictionary, value or an array, depending on 
+/// the YAML file.
+///
+/// The YAML files in the example contain objects with authors as keys,
+/// each with a sequence of their own submapping with the keys 
+/// "title" and "published"
+///
+/// ## Example
+/// ```example
+///#let bookshelf(contents) = block[
+///    #for author, works in contents [
+///        #author
+///        #list(..works.map((work => [#work.title (#work.published)])))
+///    ]
+///]
+///#bookshelf(yaml("scifi-authors.yaml"))
+/// ```
+///
+/// Display: YAML
+/// Category: data-loading
+/// Returns: array, value or dictionary
+#[func]
+pub fn yaml(
+    /// Path to a JSON file.
+    path: Spanned<EcoString>,
+) -> Value {
+    let Spanned { v: path, span } = path;
+    let path = vm.locate(&path).at(span)?;
+    let data = vm.world().file(&path).at(span)?;
+    let value: serde_yaml::Value =
+        serde_yaml::from_slice(&data).map_err(format_yaml_error).at(span)?;
+        convert_yaml(value)
+}
+
+/// Convert a YAML value to a Typst value.
+fn convert_yaml(value: serde_yaml::Value) -> Value {
+    match value {
+        serde_yaml::Value::Null => Value::None,
+        serde_yaml::Value::Bool(v) => Value::Bool(v),
+        serde_yaml::Value::Number(v) => match v.as_i64() {
+            Some(int) => Value::Int(int),
+            None => Value::Float(v.as_f64().unwrap_or(f64::NAN)),
+        },
+        serde_yaml::Value::String(v) => Value::Str(v.into()),
+        serde_yaml::Value::Sequence(v) => {
+            Value::Array(v.into_iter().map(convert_yaml).collect())
+        }
+        serde_yaml::Value::Mapping(v) => Value::Dict(
+            v.into_iter()
+                .map(|(key, value)| (convert_yaml_key(key), convert_yaml(value)))
+                .filter(|(key, _)| key.is_some())
+                .map(|(key, value)| (key.unwrap(), value))
+                .collect(),
+        )
+    }
+}
+
+/// Converts an arbitary YAML mapping key, value into a String, 
+/// or None if not possible
+fn convert_yaml_key(key: serde_yaml::Value) -> Option<Str> {
+    match key {
+        serde_yaml::Value::Null => Some(Str::from("none")),
```

Comment:
hmmm, should this be `"none"` or `"null"`? should we just disallow all non-string keys or are they common?

Reply:
Value::None gets converted to "none" by default in Typst so it should be that for the sake of concistency. But you are right that it's likely not very common and will conflict with the "none" string regardless, so we can probable sort it out.

Reply:
I think I'd prefer to just filter out all non-string keys. Typst dictionaries might gain support for non-string keys in the future and then we could add this.

Reply:
Fair enough; I'll change and document that behavior